### ISSUE TITLE
Some questions about the RMSE of the model

### ISSUE THREAD

**Issue by nowhereman1999** [sksujan58/Multivariate-time-series-forecasting-using-LSTM#3]
@sksujan58 Thanks for your great works!I have followed the article you published in the Medium to predict the Volume of stocks on my datasets successfully,However,therer are some questions about  the RMSE of the model.Here is the figure of my pred result which looks not bad yet:
![image](https://user-images.githubusercontent.com/53631206/184278635-df9e6a0e-1b9a-4025-bc25-cd8decccccb5.png)
but you can also see the RMSE is very large than I have expected and the pred at the end doesn't fit the actual value very well,it looks like "floating" above the actual value,so I wonder what's the reason of this and I have checked the formula I used to calculate the RMSE should be no problem,I would appreciate it if you can give me some advice,below is my dataset and modified code:
[new.csv](https://github.com/sksujan58/Multivariate-time-series-forecasting-using-LSTM/files/9313392/new.csv)
[timeseries Forecasting.zip](https://github.com/sksujan58/Multivariate-time-series-forecasting-using-LSTM/files/9313448/timeseries.Forecasting.zip)

**Comment by sksujan58** [comment]
Hey, 
There are something I want you to explore.
1. Please play with the hyperparameters more, I only used 8 and 10 epochs but you can increase it as much as you want, for just demo purpose I did 8,10. try to give more high epoch like 50,100,200 etc. and also play with the other parameters too.
2. your value size is very high, that's why the RMSE is also going high I guess, you can check out some other blogs too , to know about RMSE more, but high rmse does not mean that your model is bad. But please also check other error metrics. 
3. Don't stay with LSTM only, try ARIMA-SARIMAX and if you have AWS basic knowledge, try DEEPAR in sagemaker.

Hope this experiments will help you to grab more and more knowledge about timeseries. Happy learning my friend

**Comment by nowhereman1999** [comment]
> Hey, There are something I want you to explore.
> 
> 1. Please play with the hyperparameters more, I only used 8 and 10 epochs but you can increase it as much as you want, for just demo purpose I did 8,10. try to give more high epoch like 50,100,200 etc. and also play with the other parameters too.
> 2. your value size is very high, that's why the RMSE is also going high I guess, you can check out some other blogs too , to know about RMSE more, but high rmse does not mean that your model is bad. But please also check other error metrics.
> 3. Don't stay with LSTM only, try ARIMA-SARIMAX and if you have AWS basic knowledge, try DEEPAR in sagemaker.
> 
> Hope this experiments will help you to grab more and more knowledge about timeseries. Happy learning my friend

Thanks for your sincere reply!As you mentioned,I will try to give the hyperparameters include epochs a greater range and see if it improved,and about the RMSE,I think your concern is right because the value of my data is very high so the pred and origin is very high too,maybe that‘s keypoint,I will try to calculate the RMSE before the inverse transform or just find another better metrics to check the error.For the last advice,actually I have tried some transformer-based model such as informer but the result is not so good which lead me to use LSTM again,but I haven't tried on ARIMA-SARIMAX and DEEPAR yet,maybe I can have a try.Anywany,I really thanks for your detailed advice again,I will give you more feedbacks if I have made any progress!:-)

**Comment by sksujan58** [comment]
Hey, I am glad that I got the chance to help you, and one more thing, if I am not wrong, checking rmse before inverse transform , is not a good practice. It would be great if you ignore this step. Because when we did min-max we set the range of the value from 0 to 1, so in that case rmse value will be very low, and it will not give u proper visual. Try to explore more details on different metrics or graphs, to know better result.. May be you can explore more and more blogs on timeseries :)

**Comment by nowhereman1999** [comment]
> Hey, I am glad that I got the chance to help you, and one more thing, if I am not wrong, checking rmse before inverse transform , is not a good practice. It would be great if you ignore this step. Because when we did min-max we set the range of the value from 0 to 1, so in that case rmse value will be very low, and it will not give u proper visual. Try to explore more details on different metrics or graphs, to know better result.. May be you can explore more and more blogs on timeseries :)

Hi.Thanks for your timely reply,I think calculate the RMSE before inverse transform is not really a good idea after I read your analyze,so I directly try on the r2_score metrics to evaluate the model but the result is still very wired--it's a negative number as you can see in the figure:
![image](https://user-images.githubusercontent.com/53631206/184332790-4e92705e-b2ae-4f11-bdb9-271faa3137b3.png)
I'm very shocked by the result and I try to google why the RMSE of LSTM is very high and then I found a Q&A:https://datascience.stackexchange.com/questions/43263/multivariate-lstm-rmse-value-is-getting-very-high which says it is most because of the overfitting and suggest trying to start at some simple model such as GRU,but I'm wondering if it is due to my pred and origin vary widerly at peak of the origin,Anyway I really don't know how to find a suitable metrics to evaluate the model,may be I really need to explore more and more blogs on timeseries:-(

**Comment by sksujan58** [comment]
Hey, In regression by just seeing the r2_score, we can't say anything about the model. **In many blogs I saw people saying to avoid r2_score**. Please go through and check some more blogs you'll find valid reason for this may be.
By the way , don't you think your model is improved lil bit, by comparing the above two screenshots of the plots? I think it is improved lil bit. May be with more hyperparameter tuning with the all parameters you can see more positive result.
Some points I wan to add-
1. there is no fixed number for RMSE, its all depends on the target result's spread, if the value spread is higher, RMSE can go high.
2. Try to check the outliers in the target , because in many cases we have to handle outliers. It depends on the business logic.

And hey, don't get upset , ML is all about experiments, it may take long time to reach the goal. Let's enjoy the process my friend :)
Check more blogs, mine is not the ultimate solution, there are more techniques available, and I am sure you'll find it.
Happy Learning my friend

**Comment by Hezhexi2002** [comment]
> 

OK,really nice reply,as you mentioned,the model does have a little bit improvement because I have changed the range of epochs from [8,10] to [50,100] and I also reduce the variables in my dataset from 6 to 3.Actually,I'm a sophomore undergraduates from China and I'm preparing for the CUMCM-2022 right now so this is a exercise I'm currently working on,I have found another way which is decomposing the origin series first to several sub-series through EMD or VMD and then predict them independently using LSTM,afterwards,Integrating the results together.I have seen a couple of papers using this method and get great results.In the end,I'm very impressed by your opinion that you should enjoy the process,but right now I have to deal with the paper before 18 August which means I don't have much time to explore more blogs,but I will remember this during my later studying!:-)

**Comment by nowhereman1999** [comment]
> Hey, In regression by just seeing the r2_score, we can't say anything about the model. **In many blogs I saw people saying to avoid r2_score**. Please go through and check some more blogs you'll find valid reason for this may be. By the way , don't you think your model is improved lil bit, by comparing the above two screenshots of the plots? I think it is improved lil bit. May be with more hyperparameter tuning with the all parameters you can see more positive result. Some points I wan to add-
> 
> 1. there is no fixed number for RMSE, its all depends on the target result's spread, if the value spread is higher, RMSE can go high.
> 2. Try to check the outliers in the target , because in many cases we have to handle outliers. It depends on the business logic.
> 
> And hey, don't get upset , ML is all about experiments, it may take long time to reach the goal. Let's enjoy the process my friend :) Check more blogs, mine is not the ultimate solution, there are more techniques available, and I am sure you'll find it. Happy Learning my friend

OK,really nice reply,as you mentioned,the model does have a little bit improvement because I have changed the range of epochs from [8,10] to [50,100] and I also reduce the variables in my dataset from 6 to 3.Actually,I'm a sophomore undergraduates from China and I'm preparing for the CUMCM-2022 right now so this is a exercise I'm currently working on,I have found another way which is decomposing the origin series first to several sub-series through EMD or VMD and then predict them independently using LSTM.Finally,Integrating the results together.I have seen a couple of papers using this method and get great results.In the end,I'm very impressed by your opinion that you should enjoy the process,but right now I have to deal with the paper before 18 August which means I don't have much time to explore more blogs,but I will remember this during my later studying!:-)

**Comment by sksujan58** [comment]
Hey buddy, All the best for your preparation. I hope my code, and this conversation is useful to you. And timeseries is not a small topic, we have to spend months to understand it properly. But as you are out of time you have to do this quick. I am pretty sure you'll achieve your success. And one more thing to add, not only epoch, try to change the other hyperparameters too like the batch_size and also check in google, may be you can add some more new hyperparameters to play.

**Comment by nowhereman1999** [comment]
@sksujan58 Hi,sorry to bother you again,I recently use this model to predict: https://github.com/FateMurphy/CEEMDAN-VMD-GRU and get a nice result:
![image](https://user-images.githubusercontent.com/53631206/184566331-86066390-02ed-4435-bef2-a7d1865e2297.png)
you can see the r2 and MAPE are much better than before,but the RMSE and MAE are still large.However,I believe it is due to the large scale of my original data after I asked the author of CEEMDAN-VMD-GRU and he said I should calculate the RMSE and MAE with the scale of original data which he had wrote in his code,so I recomputed the metrics with scale and everything seems great now:
![image](https://user-images.githubusercontent.com/53631206/184566781-7c062427-0a8f-4cc6-b7d1-ff9f1704df11.png)
Besides,he also suggested me should focus on optimizing the r2 and MAPE because the fomula used to calculata these two metrics will be not influenced by the large scale of the original data.Anyway,now I start to use your model again because I want to use the indicators related to the Volume to adjust the result I just use Volume,but I found some problems in your code right now:
![image](https://user-images.githubusercontent.com/53631206/184567396-d1f96aaa-1f6d-474a-8b43-7adad62014e3.png)
as you can see,when I use your original code to devide the testX and testY,It will be 48 less than the df_for_training_scaled's length which I specified to 912 because I want to predict the future 912 days of  my dataset:
![image](https://user-images.githubusercontent.com/53631206/184567822-ce24da9a-7b38-4c89-9bd4-c270dffb5e22.png)
I think this is due to the createXY starts at the 48th data which causes the testX and testY lost 48 datas,and this also causes the data I predict using testX is not the future 912 datas I want to predict,instead,they are the last 912 datas of my trainX.So I change the way of division like this:
![image](https://user-images.githubusercontent.com/53631206/184568143-81509636-a4ec-4cba-9be6-759b0a1eaf40.png)
I want to use the 912 datas from df_for_training_scaled to divide testX and testY,but the shape of testX will lose the dimension of 48 and this maybe the reason of the error：
![image](https://user-images.githubusercontent.com/53631206/184569014-a80591cc-0939-420a-a2c8-0630feb51a43.png)
so I hope you can give me some advice about how to modify the testX and testY to the real future 912 datas？Best wish to you！

**Comment by sksujan58** [comment]
Brother ...you messed up with the shape..Please check the blog again. As per my blog what I did, I used previous 30 days data to predict the next day value. 
<img width="376" alt="Screenshot 2022-08-15 163201" src="https://user-images.githubusercontent.com/61758643/184624381-d63e7e35-f9fe-4ff1-a1c5-2de715dac080.png">

As per my blog , if you see the createXY function , i used 30 days as n_past to predict the next value, so in y I saved the 31st value.

**Comment by nowhereman1999** [comment]
> Brother ...you messed up with the shape..Please check the blog again. As per my blog what I did, I used previous 30 days data to predict the next day value. <img alt="Screenshot 2022-08-15 163201" width="376" src="https://user-images.githubusercontent.com/61758643/184624381-d63e7e35-f9fe-4ff1-a1c5-2de715dac080.png">

I know you use the previous 30 days to predict the next value,I change it to 48 because my it's my data's period.I know the shape should have three dimensions,but my problem is that the testX created by createXY() is not the data I want to predict,it's the last 912 data of trainX and the testY  is also not the true value of the 912 datas,the keypoint is that the df_for_testing_scaled  
is already the data I want to predict,but after the createXY the length of textX and testY all lost 48 datas which means they only have 864 datas,so I think maybe the testX and testY you divide has a little problems,it can't match the real data I want to predict

**Comment by sksujan58** [comment]
Check the shape you have in your test data- (912,5)
so when you'll pass it to createXY, -- the testX shape is -(864,48,5)
and its the correct one,  -(864+48)=912... its just the way the data is getting divided. bcoz everytime we are removing one data from the top , so total 48 times data will get removed.

**Comment by nowhereman1999** [comment]
> bcoz

Yeah,I know the createXY will remove the first 48 data，but I need to predict the full 912 data of my test data,that's what I'm stucking with now:-(

**Comment by sksujan58** [comment]
Bro, then may be you have to check with some other way, as per the blog, it is the way the division is happening, but one thing, in the final prediction we are using 30 days data to predict 31st data, you can do that too right? 48 days data to predict the 49th days value and like this go on till 912 ..by using loop

**Comment by nowhereman1999** [comment]
> Bro, then may be you have to check with some other way, as per the blog, it is the way the division is happening, but one thing, in the final prediction we are using 30 days data to predict 31st data, you can do that too right? 48 days data to predict the 49th days value and like this go on till 912 ..by using loop

Yes，I can use 48 days data to predict the 49th days value，but there are 864 data I predict instead of 912

**Comment by sksujan58** [comment]
> > Bro, then may be you have to check with some other way, as per the blog, it is the way the division is happening, but one thing, in the final prediction we are using 30 days data to predict 31st data, you can do that too right? 48 days data to predict the 49th days value and like this go on till 912 ..by using loop
> 
> Yes，I can use 48 days data to predict the 49th days value，but there are 864 data I predict instead of 912

Then run the loop again after 864,,, may be you'll get the result.. but yeah..You understood what I am saying. thats the way I did the things bro..It's just the way the data get divided.

**Comment by nowhereman1999** [comment]
> > > Bro, then may be you have to check with some other way, as per the blog, it is the way the division is happening, but one thing, in the final prediction we are using 30 days data to predict 31st data, you can do that too right? 48 days data to predict the 49th days value and like this go on till 912 ..by using loop
> > 
> > 
> > Yes，I can use 48 days data to predict the 49th days value，but there are 864 data I predict instead of 912
> 
> Then run the loop again after 864,,, may be you'll get the result.. but yeah..You understood what I am saying. thats the way I did the things bro..It's just the way the data get divided.

![image](https://user-images.githubusercontent.com/53631206/184629438-336eb5a4-83c6-44fd-9ca0-2841c4956388.png)
I try to reshape the x_test but it seems like the size of the array can't add a dimension of 48

**Comment by sksujan58** [comment]
No, bro its not possible what you are doing,,,you can't add dimension like this. You may have the check other blogs to reach the goal I guess,, bcoz this is not the correct way, you'll get errors only

**Comment by sksujan58** [comment]
But one thing, as per my blog, I used 30 days data to predict the 31st, and did it 30 times.
So you can do the same, use 48 days data to predict the 49th and do it for how long you want to do

<img width="571" alt="Screenshot 2022-08-15 172055" src="https://user-images.githubusercontent.com/61758643/184630334-11cab98b-5236-4978-bbba-fe4ab0e1f709.png">

**Comment by sksujan58** [comment]
Train test split is just to check the model accuracy..right?  so you can ignore that part,
The main thing is the final prediction, here shape does not matter.. bcoz we are doing prediction only for 1 value by using 48 past data...so do it how many times you want

**Comment by Hezhexi2002** [comment]
> But one thing, as per my blog, I used 30 days data to predict the 31st, and did it 30 times.
> So you can do the same, use 48 days data to predict the 49th and do it for how long you want to do
> 
> <img width="571" alt="Screenshot 2022-08-15 172055" src="https://user-images.githubusercontent.com/61758643/184630334-11cab98b-5236-4978-bbba-fe4ab0e1f709.png">
> 

ok,really thanks for your patience，I will check more blog to see if it can be achieved😃

**Comment by nowhereman1999** [comment]
> No, bro its not possible what you are doing,,,you can't add dimension like this. You may have the check other blogs to reach the goal I guess,, bcoz this is not the correct way, you'll get errors only

ok,really thanks for your patience，I will check more blog to see if it can be achieved😃

**Comment by nowhereman1999** [comment]
> But one thing, as per my blog, I used 30 days data to predict the 31st, and did it 30 times.
> So you can do the same, use 48 days data to predict the 49th and do it for how long you want to do
> 
> <img width="571" alt="Screenshot 2022-08-15 172055" src="https://user-images.githubusercontent.com/61758643/184630334-11cab98b-5236-4978-bbba-fe4ab0e1f709.png">
> 

ok，I will try this idea later and see if it works

**Comment by sksujan58** [comment]
> > No, bro its not possible what you are doing,,,you can't add dimension like this. You may have the check other blogs to reach the goal I guess,, bcoz this is not the correct way, you'll get errors only
> 
> ok,really thanks for your patience，I will check more blog to see if it can be achieved😃

No need to thank me bro.. I just want to clear my thoughts to you and want you to get the success... and that's the way it works. All the best
